security/src/main/java/pm/gnosis/heimdall/security/impls/CompatSafeFingerprintManager.kt

```
@@ -0,0 +1,130 @@
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun createKey(): SecretKey {
+        val builder = KeyGenParameterSpec.Builder(FINGERPRINT_KEY,
+                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
+                .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                .setUserAuthenticationRequired(true)
+                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+        keyGenerator.init(builder.setKeySize(256).build())
+        return keyGenerator.generateKey()
+    }
+
+    fun removeKey(): Completable = Completable.fromCallable {
+        keyStore.load(null)
+        keyStore.deleteEntry(FINGERPRINT_KEY)
+    }.subscribeOn(Schedulers.io())
+
+    fun systemHasFingerprintsEnrolled() =
+            nullOnThrow { FingerprintManagerCompat.from(context).hasEnrolledFingerprints() } ?: false
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun getOrCreateKey(): SecretKey {
+        keyStore.load(null)
+        return keyStore.getKey(FINGERPRINT_KEY, null) as? SecretKey ?: createKey()
+    }
+
+    fun isKeySet(): Single<Boolean> = Single.fromCallable {
+        keyStore.load(null)
+        keyStore.getKey(FINGERPRINT_KEY, null) != null
+    }.subscribeOn(Schedulers.io())
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun createCipher(iv: ByteArray?) =
+            Cipher.getInstance(KeyProperties.KEY_ALGORITHM_AES + "/"
+                    + KeyProperties.BLOCK_MODE_CBC + "/"
+                    + KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                    .apply {
+                        if (iv == null) init(Cipher.ENCRYPT_MODE, getOrCreateKey())
+                        else init(Cipher.DECRYPT_MODE, getOrCreateKey(), IvParameterSpec(iv))
+                    }
+
+    fun authenticate(iv: ByteArray? = null): Observable<AuthenticationResult> =
+            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) Observable.empty()
```

We could also throw an exception here